rfcs/2021-12-01-9460-healthcheck-endpoint-improvements

```diff
@@ -0,0 +1,114 @@
+# RFC 9460 - 2021-12-01 - Healthcheck endpoint improvements
+
+Our existing health endpoint is too limited and additional features are required
+to properly interact with load balancers or Kubernetes service discovery.
+
+## Context
+
+- [Additional response for api health endpoint](https://github.com/vectordotdev/vector/issues/9160)
+
+## Scope
+
+### In scope
+
+- Update health endpoint to improve Vector's reliability in production deployments
+- Update default configurations to improve Vector's out-of-the-box reliability
+
+### Out of scope
+
+- Moving the health endpoint out of our existing API
+- Adding user facing configuration to adjust endpoint behavior or responses
+
+## Pain
+
+- Today's health endpoint causes 502's when running behind a loadbalancer
+- The healthcheck isn't integrated with any components and only represents if Vector
+itself is running
+
+## Proposal
+
+### User Experience
+
+The existing health endpoint on our GraphQL based API is enhanced to return 503's
+when Vector is shutting down. This additional response will allow load balancers
+and service discovery to better integrate with Vector, sources already begin
+rejecting requests received during the shutdown process and updating the health
+check will allow for removing the instance from available backends and avoid
+routing traffic to an instance that will end up rejecting the request regardless.
+
+### Implementation
+
+We have existing logic in place for other components to handle Vector's shutdown.
+The [health handler](https://github.com/vectordotdev/vector/blob/master/src/api/handler.rs#L7)
+can be updated to respond 503 if shutdown has started and 200 otherwise. This
+should be a minor change and not negatively impact existing deployments.
+
+## Rationale
+
+This improvement is low effort but greatly improves the experience of running
+Vector behind a load balancer or in a service discovery system. As we reccomend
+both of those options for production deployments there's very little reason not
+to implement this.
+
+## Drawbacks
+
+This is a minor change overall and does not add any real engineering burden.
+It does open the door for further health check improvements which are more
+complex. Any further improvements to the check need to be planned and discussed
+to avoid unwanted complexity.
+
+## Prior Art
+
+- [Beats](https://www.elastic.co/guide/en/beats/filebeat/7.15/http-endpoint.html)
+  - [Liveness and Readiness Probes](https://github.com/elastic/helm-charts/blob/main/filebeat/values.yaml#L174-L198)
+- [FluentBit](https://docs.fluentbit.io/manual/administration/monitoring#rest-api-interface)
+  - [Liveness and Readiness Probes](https://github.com/fluent/helm-charts/blob/main/charts/fluent-bit/values.yaml#L131-L139)
```

Comment:
Very small nit, it's best to link to specific SHAs to ensure the code is always what you meant to link to. The `fluent-bit` one has already been updated in a way that it is one line off.

Reply:
Resolved in ef8a799